zephyr: a few quality-of-life improvements, and additional board configurations

dpgeorge opened the pull request:
### Summary

The main aim here was to get `mpremote mip install unittest` working on the zephyr port, so it could run more of the test suite.  And that lead to a lot of little changes.

Overall this PR aims to bring the zephyr port a bit closer to other ports in terms of functionality and usability:
- execute `boot.py` and `main.py` like other ports
- add `/flash/lib` to `sys.path` on start up if there's a filesystem (eg `mpremote mip install` now works)
- enable `sys.stdin/out/err`
- improve REPL reliability and performance, it now passes the test from #15909, and copying files with `mpremote` is now reliable
- fix USB device_next driver compilation error
- add `Pin.OPEN_DRAIN` support
- allow constructing a Pin from an existing Pin (follows other ports)
- implement C-level pin HAL
- enable `machine.SoftI2C`
- enable importing of .mpy files

It also tweaks/adds a few board configurations:
- frdm_k64f: get ctrl-C working, enable PWM, increase heap size
- nucleo_wb55rg: enable BLE, I2C, SPI and add filesystem
- rpi_pico: use USB for REPL, enable I2C, SPI, add filesystem, increase heap size

In particular, the filesystem settings for `rpi_pico` now match the rp2 port RPI_PICO board, so you can switch between zephyr and rp2 firmware and retain the filesystem (and even .mpy files on that filesystem work across the two ports!).  That allows some nice side-by-side comparisons between the zephyr and rp2 ports for the same hardware.  Eg USB CDC throughput is quite a lot higher on the rp2 port using TinyUSB, compared with zephyr's USB device (but maybe this is due to zephyr configuration issues, I didn't investigate further).

### Testing

Tested with frdm_k64f, nucleo_wb55rg and rpi_pico zephyr firmware:
- The serial test from #15909 now passes on all three boards.
- Can do `mpremote mip install unittest` and it works.
- Can now run a lot more of the test suite, including `--via-mpy` tests.
- Tested `machine.Pin.OPEN_DRAIN` and `machine.SoftI2C` (actually as part of #17365).

### Trade-offs and Alternatives

This increases the firmware size of zephyr builds a bit, but (1) these boards have enough flash and RAM; and (2) it's easy enough for custom boards to reconfigure/disable features if needed.

dpgeorge committed:
zephyr/boards/rpi_pico: Add board configuration for rpi_pico.

Although the rpi_pico can already build and run with the zephyr port, this
configuration improves it in a number of ways:
- Use the USB CDC ACM as the REPL, rather than just a UART.
- Enable I2C and SPI, and add I2C1.
- Enable a filesystem, which matches exactly the rp2 port's RPI_PICO
  configuration.  So switching between zephyr and rp2 is possible and will
  retain the filesystem.
- Make the MicroPython GC heap make the most use of the available RAM.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
zephyr/main: Execute boot.py and main.py like other ports.

Changes here make the zephyr port act the same as other ports for the
start up and shut down sequence:
- `boot.py` is executed if it exists, and can force a soft reset
- `main.py` is only executed if in friendly REPL and if `boot.py` executed
  successfully; and it can also force a soft reset
- print "MPY: " before "soft reboot" on soft reset

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
zephyr/main: Add /flash/lib or /sd/lib to sys.path on start up.

If there is a filesystem available, this change makes sure there is a "lib"
in `sys.path`, eg so that "mip install" works correctly.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
zephyr: Enable sys.stdin/out/err.

This change enables `sys.stdin`, `sys.stdout` and `sys.stderr` objects.
They are useful for general IO, and also help with testing zephyr boards.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
zephyr/src: Increase UART input buffer to 512 bytes and reduce latency.

There are two changes here:

1. Increase the UART input bufffer to 512 bytes.  That's necessary to get
   basic REPL reliability tests working, and helps improve `mpremote`
   usage, eg copying large files.

2. Remove `uart_sem` semaphore.  This is no longer needed because
   `zephyr_getchar()` should be fully non-blocking and have as low a
   latency as possible.  `mp_hal_stdin_rx_chr()` (which calls
   `zephyr_getchar`) already uses `MICROPY_EVENT_POLL_HOOK` to get
   an efficient wait, and doing an extra wait and check for the
   semaphore in `zephyr_getchar()` just introduces unnecessary latency and
   can lead to slower input, and potentially overflowing the UART input
   buffer.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
zephyr/src: Fix USB device_next driver to work with zephyr 4.0.0.

The blocklist argument is not available in zephyr 4.0.0.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
zephyr/machine_pin: Add Pin.OPEN_DRAIN constant.

Adding this constant is all that's needed to support open-drain pins.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
zephyr/machine_pin: Allow constructing a Pin with an existing Pin.

Following other ports.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
zephyr/mphalport: Implement C-level pin HAL.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
zephyr/mpconfigport: Enable machine.SoftI2C and machine.SoftSPI.

These work now that the C-level pin HAL is implemented.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
zephyr/mpconfigport: Enable import of mpy and a few related features.

Support for importing .mpy files is quite fundamental to MicroPython these
days, eg it allows installing more efficient .mpy code via "mip install"
(and installing `unittest` only works with the .mpy version because the .py
version uses f-strings, which are not enabled on the zephyr port).  So
enable it generally for use by all boards.

As part of this, also enable:
- min/max: needed by `micropython/import_mpy_invalid.py`, and widely used
- sys.modules: needed by `run-tests.py` to run .mpy tests with --via-mpy
- io module: needed to run .mpy tests, and useful for `io.IOBase`
- array slice assign: needed to run .mpy tests, and generally useful as a
  way to do a memory copy.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
zephyr/mpconfigport: Enable MICROPY_NLR_THUMB_USE_LONG_JUMP.

Needed for some ARMv6M boards, eg rpi_pico.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
zephyr/boards/frdm_k64f: Improve board configuration.

Changes:
- Enable CONFIG_PWM so that `machine.PWM()` works.
- Increase MicroPython GC heap size.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
zephyr/boards/nucleo_wb55rg: Enable BLE, I2C, SPI and add filesystem.

Bluetooth works well now on this board, so enable all supported features.

Also increase the MicroPython GC heap size to make use of the available
RAM.

Unfortunately the filesystem does not match the stm32 port's NUCLEO_WB55
configuration.  That's not possible to do because stm32 uses a 512 byte
flash erase size, while zephyr uses 4096 bytes.  But at least here in
zephyr there's now a sizable and usable filesystem.

Signed-off-by: Damien George <damien@micropython.org>

bikeNomad commented on ports/zephyr/machine_pin.c:
```diff
@@ -271,6 +279,7 @@ static const mp_rom_map_elem_t machine_pin_locals_dict_table[] = {
     // class constants
     { MP_ROM_QSTR(MP_QSTR_IN),        MP_ROM_INT(GPIO_INPUT) },
     { MP_ROM_QSTR(MP_QSTR_OUT),       MP_ROM_INT(GPIO_OUTPUT) },
+    { MP_ROM_QSTR(MP_QSTR_OPEN_DRAIN), MP_ROM_INT(GPIO_OUTPUT | GPIO_OPEN_DRAIN) },
```
Should this be:
```c
{ MP_ROM_QSTR(MP_QSTR_OPEN_DRAIN), MP_ROM_INT(GPIO_OUTPUT | GPIO_OPEN_DRAIN | GPIO_INPUT) },
```

dpgeorge replied:
Done.

bikeNomad commented on ports/zephyr/mphalport.h:
```diff
@@ -35,3 +36,33 @@ static inline uint64_t mp_hal_time_ns(void) {
+
+static inline void mp_hal_pin_open_drain(mp_hal_pin_obj_t pin) {
+    (void)gpio_pin_configure(pin->port, pin->pin, GPIO_OUTPUT | GPIO_OPEN_DRAIN);
+}
```
Should this be:
```c
    (void)gpio_pin_configure(pin->port, pin->pin, GPIO_OUTPUT | GPIO_OPEN_DRAIN | GPIO_INPUT);
```

dpgeorge replied:
Done.

bikeNomad reviewed (changes requested):
Just a couple of comments regarding being able to read from output/open-drain Pins.

dpgeorge commented on ports/zephyr/src/usbd.c:
```diff
@@ -137,7 +131,7 @@ struct usbd_context *mp_usbd_init_device(usbd_msg_cb_t msg_cb) {
     }
 
-    err = usbd_register_all_classes(&mp_usbd, USBD_SPEED_FS, 1, blocklist);
+    err = usbd_register_all_classes(&mp_usbd, USBD_SPEED_FS, 1);
```
I'm using 4.0.0, which definitely has only 3 parameters... I guess we need to have a `#if` here to support that and newer versions at the same time?

dpgeorge replied:
OK, they added the blocklist in https://github.com/zephyrproject-rtos/zephyr/commit/3bfe16381157019a70c57d1e363172decad25949, which is in v4.1.0 but not v4.0.0.

dpgeorge committed:
zephyr/mpconfigport: Enable sys.maxsize.

Costs +48 bytes.  Useful to introspect the target.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
zephyr/mpconfigport: Enable emergency exception buffer.

Needed to pass exception tests.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
zephyr/machine_timer: Make machine.Timer id argument optional.

With a default of -1, for soft timer.  This matches other ports, and the
`extmod/machine_timer.c` implementation.

This change allows the `tests/extmod/machine_soft_timer.py` test to pass.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
I've added 3 more commits/changes here:
- enable `sys.maxsize`; that costs +48 bytes but allows introspection to the target, needed at least for some tests to work
- enable emergency exception buffer, again needed for a few tests to run and pass (otherwise the device can lock up running the `tests/micropython/heapalloc_exc_compressed_emg_exc.py` test)
- make the id argument to `machine.Timer` optional, defaulting to -1 (they only allowed value); that matches other ports and gets the `tests/extmod/machine_soft_timer.py` test passing

Testing was done with `rpi_pico` firmware.

dpgeorge committed:
zephyr/machine_pin: Retry configuring gpio with just GPIO_OUTPUT.

Some targets like frdm_k64f don't support GPIO_OUTPUT|GPIO_INPUT, so just
use GPIO_OUTPUT in those cases (it seems they still support reading the
current output state even when configured only as GPIO_OUTPUT, unlike other
targets which require both settings).

Signed-off-by: Damien George <damien@micropython.org>

jonnor commented:
Very nice! Zephyr support seems to be soaring ahead lately :D
